### PR TITLE
Calendar categories

### PR DESCRIPTION
**Pull request by nquinn721**
## Description
Updated the category view to handle styles within each interval based off of category. I also added the ability to take an object in the category list and you can specify which prop to display as the text for category through 'category-text'. 'category-text' can be a string or a function where you can run logic to determine what to show.

## Motivation and Context
Our app needed to color each interval in the calendar based off of category as well as we wanted meta data in the category. 

## How Has This Been Tested?
I've updated the unit tests to include my changes

## Markup:

<details>

```vue
<template>
  <v-container>
    <v-calendar
      color="primary"
      type="category"
      :categories="[{ name: 'Jeff' }, { name: 'Nate' }]"
      category-text="name"
      category-show-all
      :events="events"
    />
  </v-container>
</template>

<script>
export default {
  data: () => ({
    events: [
      {
        name: "Nate",
        category: "Nate",
        start: new Date(),
        end: new Date()
      }
    ]
  })
};
</script>

```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [x] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

**Comment by jacekkarczmarczyk** [comment]
I'm having same error as in previous PR:

image

There are also some unrelated changes, for example in `packages/docs/src/layouts/default/ListItem.vue` file

**Comment by nquinn721** [comment]
> I'm having same error as in previous PR:
> 
> image
> 
> There are also some unrelated changes, for example in `packages/docs/src/layouts/default/ListItem.vue` file

What is your playground that you are getting this error?

**Comment by jacekkarczmarczyk** [comment]
The one that's in the PR description

**Comment by nquinn721** [comment]
> The one that's in the PR description

Sorry, i missed some stuff. I've updated it now

File: packages/vuetify/src/components/VCalendar/VCalendar.ts [hunk 1]
```diff
@@ -193,6 +198,11 @@ export default CalendarWithEvents.extend({
   },
 
   methods: {
+    parsedCategoryText (category: any): string {
+      return typeof this.categoryText === 'string'
+        ? category[this.categoryText as string]
+        : this.categoryText(category) as string
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendar.ts:204]:
```suggestion
        ? category[this.categoryText]
        : this.categoryText(category)
```

File: packages/vuetify/src/components/VCalendar/VCalendar.ts [hunk 2]
```diff
@@ -335,9 +345,10 @@ export default CalendarWithEvents.extend({
           }
         }
 
-        categories = Object.keys(categoryMap)
+        categories = categories.filter(v => {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendar.ts:348]: can be one-lined

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts [hunk 3]
```diff
@@ -30,12 +30,21 @@ export default VCalendarDaily.extend({
       return typeof this.categories === 'string' && this.categories
         ? this.categories.split(/\s*,\s*/)
         : Array.isArray(this.categories)
-          ? this.categories as string[]
+          ? this.categories.map((v: any) => {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:33]: code is duplicated in Calendar.ts, could be extracted to an util along with parseCategoryText

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts [hunk 4]
```diff
@@ -30,12 +30,21 @@ export default VCalendarDaily.extend({
...
+            const categoryName = this.parsedCategoryText(v)
+            if (typeof v === 'string') v = { name: v }
+            v.categoryName = categoryName
+            return v
+          }) as string[]
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:38]: it doesn't seem to return `string[]`

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts [hunk 5]
```diff
@@ -44,14 +53,16 @@ export default VCalendarDaily.extend({
...
 
-      const children = this.parsedCategories.map(category => this.genDayHeaderCategory(day, this.getCategoryScope(scope, category)))
+      const children = this.parsedCategories.map(category => {
+        return this.genDayHeaderCategory(day, this.getCategoryScope(scope, category))
+      })
 
       return [this.$createElement('div', data, children)]
     },
-    getCategoryScope (scope: any, category: string) {
+    getCategoryScope (scope: any, category: any) {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:62]:
```suggestion
    getCategoryScope (scope: any, category: { categoryName: string }) {
```

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts [hunk 6]
```diff
@@ -61,25 +72,69 @@ export default VCalendarDaily.extend({
           return this.getCategoryScope(this.getSlotScope(day), scope.category)
         }),
       }, [
-        getSlot(this, 'category', scope) || this.genDayHeaderCategoryTitle(scope.category),
+        getSlot(this, 'category', scope) || this.genDayHeaderCategoryTitle(scope.category && scope.category.categoryName),
         getSlot(this, 'day-header', scope),
       ])
     },
-    genDayHeaderCategoryTitle (category: string) {
+    genDayHeaderCategoryTitle (categoryName: any) {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:79]:
```suggestion
    genDayHeaderCategoryTitle (categoryName: string | null) {
```
(or whatever type it takes)

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts [hunk 7]
```diff
@@ -61,25 +72,69 @@ export default VCalendarDaily.extend({
...
       }
 
-      const children = this.parsedCategories.map(category => this.genDayBodyCategory(day, category))
+      // const children = this.parsedCategories.map(category => this.genDayBodyCategory(day, category))
+      const children = [this.genDayBodyCategory(day, category)]
 
       return [this.$createElement('div', data, children)]
     },
-    genDayBodyCategory (day: CalendarTimestamp, category: string): VNode {
+    genDayBodyCategory (day: CalendarTimestamp, category: any): VNode {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:137]:
```suggestion
    genDayBodyCategory (day: CalendarTimestamp, category: { categoryName: string }): VNode {
```

File: packages/vuetify/src/components/VCalendar/VCalendarCategory.ts (hunk 5 above)
Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VCalendar/VCalendarCategory.ts:62]: typing should be fixed in similar manner in other places too, also why not passing just the string instead of the object with `categoryName` prop?

File: packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts [hunk 8]
```diff
@@ -126,11 +125,17 @@ export default CalendarBase.extend({
       return this.parsedWeekdays
     },
     categoryMode (): boolean {
-      return false
+      return this.type === 'category'
     },
   },
 
   methods: {
+    eventColorFunction (e: CalendarEvent): string {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/components/VCalendar/mixins/calendar-with-events.ts:133]: how is this change related to categories?

also `e.color ? e.color : this.eventColor` could be simplified to `e.color || this.eventColor`

**Review by jacekkarczmarczyk — CHANGES_REQUESTED** [review]
it's a breaking change, the following code works before this PR, but throws an error with this PR

```html
    <v-calendar
      color="primary"
      type="category"
      :categories="['Jeff', 'Nate']"
      category-show-all
      :events="events"
    />
```

Also please remove unrelated commits

File: packages/vuetify/src/components/VCalendar/VCalendar.ts (hunk 2 above)
Comment (nquinn721) [reply packages/vuetify/src/components/VCalendar/VCalendar.ts:348]: Not sure I understand this one

Comment (jacekkarczmarczyk) [reply packages/vuetify/src/components/VCalendar/VCalendar.ts:348]: `categories = categories.filter(v => Object.keys(categoryMap).includes(v.categoryName))`

**Comment by nquinn721** [comment]
@jacekkarczmarczyk I fixed the break in my change

**Comment by nekosaur** [comment]
You still have a number of unrelated commits in this PR. And you're still missing documentation changes

**Comment by nquinn721** [comment]
> You still have a number of unrelated commits in this PR. And you're still missing documentation changes

So i think these commits came from master since I merged in with master first. I had documentation. Maybe the new style of documentation got rid of mine.

**Comment by nquinn721** [comment]
@nekosaur  Looks like the documentation is already up to date to cover the changes

**Comment by nekosaur** [comment]
In what way? It was my understanding that you've at the very least added a new prop

**Comment by nquinn721** [comment]
@nekosaur Sorry I realized what I was doing wrong. I was looking at docs/src/api and I saw that 'category-text' was in there but it was not in the v-calendar.json file. I just added it.

**Comment by nquinn721** [comment]
@nekosaur do i need to solve the other commits, honestly I'm not sure how to handle those.

**Comment by nekosaur** [comment]
Yes, you need to rebase the branch so that it only contains your commits

**Comment by nquinn721** [comment]
@nekosaur @jacekkarczmarczyk here's my new pr https://github.com/vuetifyjs/vuetify/pull/12518

**Comment by KaelWD** [comment]
You should just force push instead of creating a new PR, too late now though.
